rfcs/0000-shareable-courses.md

```diff
@@ -0,0 +1,150 @@
+
+# Detailed design
+
+Shareable courses will be modeled identically to normal courses: as GitHub repositories containing questions.
+
+## Usage across multiple PrairieLearn instances.
+
+By design, PrairieLearn is a decentralized system. There are now several production instances running at different universities, and dozens (if not hundreds) of individuals run PrairieLearn on their computers for local course development. The architecture of shareable courses needs to take this decentralized nature into account.
+
+Making a shareable course available to an institution's instance of PrairieLearn will be as simple as syncing it like a normal course. Updates to shareable courses are also handled via syncing.
+
+For the first iteration of shareable courses, syncing to acquire updates will be a manual process. In the future, a syncing protocol could be designed so that shareable courses are automatically updated across all production PrairieLearn instances where they're in use.
+
+For local development, there are two scenarios:
+
+* The person running PrairieLearn locally does not have the necessary permissions to clone the shareable course repository, or they do but have not previously cloned it. In that case, an attempt to access an external question from a shareable course will provide a clear error page explaining why the question cannot be viewed there and offering instructions on how to clone and add the course.
+* The person running PrairieLearn locally has cloned the shareable course repository. In this case, they can add the course to their local Docker container. Then, questions from that shareable course can be accessed exactly like they would be in a production environment.
+
+In the future, one could build a feature that allows an authorized user to export a copy of a shareable course for use with local development.
+
+## Safety guarantees
+
+The following guarantees should exist:
+
+* There is no ambiguity between whether a question belongs to a course or an external shareable course.
+* Syncing a new version of a shareable course should not break any existing assessments that reference questions from it.
+* There should never be an assessment on a production instance that references a question that does not exist.
+
+The following facts and behaviors ensure those guarantees:
+
+* A leading `@` in a question ID in an assessment definition will unambiguously differentiate shareable course questions from course-local questions.
+
+* Once a course has been designated as shareable, it must remain shareable. That is, once a course with `"shareable": true` is synced to an instance of PrairieLearn, attempting to sync the same course with `"shareable": false` will be a sync error.
+
+* Once a question in a shareable course has been synced to an instance of PrairieLearn, attempting to perform a sync where the question would be renamed or deleted will be a sync error. *(Is there any special consideration necessary for local development?)*
```

Comment:
I think that in local development all of these safety guarantees should be disabled, so that you can rename new questions before pushing them to production.